[FEATURE] 유저, 인증 기능 구현

JakePark929 opened the pull request:
## 📝 PR Summary

유저 회원 가입 및 로그인 구글 OAuth 2.0 + jwt 이용해서 구현했습니다.

- User 도메인이 추가되었습니다.
- Jwt 라이브러리를 추가했습니다.
- Google OAuth2.0 + Jwt 로그인기능을 구현하였습니다.
 > 로그인 인증이 완료되면 구글 API 로 부터 받은 유저정보를 기반으로 Jwt 발행을 진행합니다.
 > 만약 회원가입되어 있지 않은 유저라면 회원가입을 자동으로 진행합니다.
- RefreshToken 기반으로 새로운 AccessToken 을 발급하는 API 를 추가했습니다.
- User 닉네임 수정 API 를 추가했습니다.

#### 🌲 Working Branch

feat/#6-oauth

#### 🌲 TODOs

- [x] DB 생성 - 우선 로컬에 Entity 만 생성하여 진행예정
- [x] Google OAuth 회원 가입 진행
- [x] 회원 닉네임 수정

### Related Issues

This resolve #6 

[BE] [FEATURE] 유저, 인증 기능 구현, @JakePark929 

#### 📚 Remarks
- 회원 탈퇴는 추후에 별도 이슈로 등록하여 진행 하겠습니다.
- 카카오톡 로그인도 같이 추가하려 했으나 초반부터 분량이 너무 많은 것 같아서 우선 구글 로그인만 진행했습니다.
  > 바로 추가할 수 있도록 결합도를 최대한 낮췄습니다. 별도 이슈 등록하여 처리하겠습니다.
- jwt secret키와 google api 키는 별도 채널로 공유하겠습니다.
- Swagger 를 등록해보려고 했는데 템플릿 설정 문제인지 적용이 되지 않습니다. 별도 이슈로 등록하여 처리 진행하겠습니다.
- 현재 구글 API 에서 생년월일, 성별 등과 같은 정보를 추가하려면 더 많은 설정이 필요해서 우선 null 로 처리합니다. 중간에 회원가입 같은 로직을 클라이언트와 함께 추가해야 될 것 같습니다. 이 부분도 미리 이슈로 등록하겠습니다.

이상 다들 리뷰 부탁드려요~

JakePark929 committed:
chore : jwt 라이브러리 추가, jpa 설정 변경 및 키 추가

JakePark929 committed:
[#6] feat : User 도메인 추가 및 Google OAuth2.0 + JWT 구현

- User 도메인이 추가되었습니다.
- Jwt 라이브러리를 추가했습니다.
- Google OAuth2.0 + Jwt 로그인기능을 구현하였습니다.
 > 로그인 인증이 완료되면 구글 API 로 부터 받은 유저정보를 기반으로 Jwt 발행을 진행합니다.
- RefreshToken 기반으로 새로운 AccessToken 을 발급하는 API 를 추가했습니다.
- User 닉네임 수정 API 를 추가했습니다.

JakePark929 committed:
style : 불필요한 공백 제거

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/client/OAuth2Client.java:
```diff
@@ -0,0 +1,59 @@
+
+@Component
+public class OAuth2Client {
+    private final static String AUTHORIZATION_URI = "https://accounts.google.com/o/oauth2/auth";
+    private final static String TOKEN_URI = "https://oauth2.googleapis.com/token";
+    private final static String USER_INFO_URI = "https://www.googleapis.com/oauth2/v3/userinfo";
```
해당 부분은 Google에 특화되어있는 URL로 알고 있는데요.

해당 값들을 따로 주입받아 사용해 보는 것은 어떨까요.
또한 Google URL이 있으니, 추후 수정시 이해하는데 불편함은 없겠지만, 확실하게 GOOGLE의 변수라는것 정도는 인지해야 하지 않을까 하는 제 개인적인 생각입니당 :)

JakePark929 replied:
core-api yml에서 별도로 주입하도록 수정 완료 했습니다.
변수 명에도 google 넣어 줬네요 :)

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/client/OAuth2Client.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public Map<String, String> getAccessToken(String authorizationCode) {
+        RestTemplate restTemplate = new RestTemplate();
```
로그인시마다 RestTemplate를 매번 새로운 객체를 생성하는 것에 있어서, 어떤 문제점이 있을지 생각해보면 좋을 것 같습니다. :)

(저도 과거에 들었던 말이에용 😃)

JakePark929 replied:
코멘트 감사합니다! 별도 Bean 으로 수정했습니다~

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/client/OAuth2Client.java:
```diff
@@ -0,0 +1,59 @@
+
+    public Map<String, String> getAccessToken(String authorizationCode) {
+        RestTemplate restTemplate = new RestTemplate();
+        Map<String, String> request = new HashMap<>();
+
+        request.put("client_id", clientId);
+        request.put("client_secret", clientSecret);
+        request.put("redirect_uri", redirectUri);
+        request.put("grant_type", "authorization_code");
+        request.put("code", authorizationCode);
```
해당 Map도 왠지 Bean, 또는 Method로 분리가 가능한 냄새가 솔솔 나는것 같습니다.

JakePark929 replied:
Request Dto 를 만들어 분리했습니다~

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/resolver/UserSessionResolver.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        var annotation = parameter.hasParameterAnnotation(UserSession.class);
+        var parameterType = parameter.getParameterType().equals(UserDto.class);
```
프로젝트를 처음하는 분들도 있다보니, var, record등 JDK11이후의 문법은 일단 지양하고 JDK8까지들의 주 문법들을 사용해보는게 어떨까요.

저도 최대한 이해하기 쉽게 JDK8,9까지만 추가된 기능으로만 작성해볼까 합니다 :)

JakePark929 replied:
정신없이 작성하다 var 를 그대로 써버렸네요 🤣
수정 했습니다.

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/user/converter/UserConverter.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Optional;
+
+@Converter
+public class UserConverter {
+    public User toEntity(UserRegisterRequest request) {
+        return Optional.ofNullable(request)
+                .map(it -> User.builder()
+                        .subId(request.getSubId())
+                        .email(request.getEmail())
+                        .name(request.getName())
+                        .thumbnail(request.getThumbnail())
+                        .role(UserRole.USER)
+                        .status(UserStatus.ACTIVE)
+                        .build())
+                .orElseThrow(NullPointerException::new);
+    }
+
+    public UserDto toDto(User entity) {
+
+        return UserDto.of(
+                entity.getId(),
+                entity.getSubId(),
+                entity.getEmail(),
+                entity.getName(),
+                entity.getNickname(),
+                entity.getBirth(),
+                entity.getGender(),
+                entity.getThumbnail(),
+                entity.getRole(),
+                entity.getStatus(),
+                entity.getCreatedAt(),
+                entity.getUpdatedAt()
+        );
+    }
```
👍👍사랑합니다 중언님

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/annotation/Business.java:
```diff
@@ -0,0 +1,17 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Service
+public @interface Business {
```
해당 어노테이션의 사용 용도는 어떻게 되나용?

JakePark929 replied:
Business 어노테이션은 각 도메인 Business 클래스에 명시적으로 적용합니다.
entity 객체는 Service 클래스에서만 다루도록 하여 entity 객체의 전파를 막고, 추가 비즈니스적인 코드들을 Business 클래스가 다루도록 합니다. 동작은 Service 어노테이션과 동일합니다!

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/annotation/Converter.java:
```diff
@@ -0,0 +1,17 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Service
+public @interface Converter {
```
해당 어노테이션의 용도는 Entity -> Request, Request -> Entity로직만을 담당하는 Bean을 만들고자 할때 사용해서 명시하는게 목적일까요?

JakePark929 replied:
Entity와 Dto 객체의 변환용 클래스 맞습니다! Converter 클래스에 명시적으로 붙여줍니다.
동작은 Service 어노테이션과 같습니다.

donsonioc2010 reviewed (commented):
같이하자고 하셨는데 먼저 훅나가버리셔서 당황했습니다 🤣

작업량이 정말많네요 감사합니다. 집히는곳들이 있어서, Comment로해놓겠습니다 :) 한번 확인후 다시 Approve하는걸로 해야할 것 같습니다.

JakePark929 committed:
refactor : URL 외부 주입, RestTemplate 빈 생성, Request DTO 생성 외

- Google OAuth2.0 전용 URL 은 yml 파일에서 주입받도록 분리
- `getAccessToken` 메소드 Request, Response DTO 생성
- `UserSessionResolver` NPE 처리, var 변경

JakePark929 commented:
너무 끌면 클라이언트 환경 세팅이 늦어 질 거 같아
로그인 쪽 먼저 잡자 했는데 생각보다 분량이 많아서 저도 당황했네요.. 😂😂 
리뷰 사항 반영했으니 확인 해주시면 감사하겠습니다~

BellaChooi commented:
저도 찬찬히 분석하고 코멘트 달아놓겠습니다. 고생하셨습니다 :)

BellaChooi commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/annotation/UserSession.java:
```diff
@@ -0,0 +1,11 @@
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.PARAMETER)
```
해당 어노테이션(UserSession)이 적용될 수 있는 자바 요소의 종류를 지정하는것. 

ex) public String getUser(@UserSession UserSessionObject userSession) {
        return userSession.getUserName();
     }

BellaChooi commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/client/OAuth2Client.java:
```diff
@@ -0,0 +1,69 @@
+public class OAuth2Client {
+    private final RestTemplate restTemplate;
+
+    @Value("${oauth2.google.authorization-uri}")
```
스프링 프레임워크에서 설정 파일(application.properties 또는 application.yml)에 정의된 값을 필드에 주입하는 데 사용

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/_common/intercepter/AuthorizationInterceptor.java:
```diff
@@ -0,0 +1,63 @@
+            return true;
+        }
+
+        final String accessToken = request.getHeader("authorization-token");
```
OAuth 2.0 에서 토큰을 어떤 방식으로 헤더에 담는지 학습해보시면 좋을 것 같아요 : )

JakePark929 replied:
보통 JWT는 Autorization 헤더에 Bearer 를 설정하여 사용하는데요!
제가 쓰는 크롬 확장 프로그램인 ModHeader 에서 Autorization 필드로 넣었더니 서버에서 인식이 안 되더라구요.
피드백 주신 사항 듣고 생각해보니 Bearer 에 대한 설정을 안 해줘서 그럴 수도 있겠다는 생각이 듭니다!
한 번 확인해보겠습니다~

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/auth/controller/AuthOpenApiController.java:
```diff
@@ -0,0 +1,34 @@
+        return authBusiness.login();
+    }
+
+    @GetMapping("/login/oauth2/code/google")
+    public ApiResponse<JwtResponse> callbackGoogle(@RequestParam("code") String authorizationCode) {
+        final JwtResponse jwtResponse = authBusiness.callbackGoogle(authorizationCode);
+
+        return ApiResponse.success(jwtResponse);
+    }
```
Http GET으로 민감한 정보를 전달한다면 위험할 수 있다고 생각하는데 어떻게 생각하시나요?

JakePark929 replied:
우선 google 로그인 api 의 콜백 은 get 으로 리다이렉트를 해주는 것으로 알고 있는데요!
post 로 받으려면 뭔가 추가로 설정해줘야 한다고 합니다.

그 외, 별도 토큰 발행용 api 를 post 맵핑으로 추가해서 콜백 요청이 성공하면 인증 번호를 줘서 다시 서버에 발급 받는 방법,
아니면 반환 데이터 자체에 RSA 등 암호화를 하는 방법 등이 생각납니다.

혹시 다른 방법이 있으면 알려 주시면 감사하겠습니다! :)
